crates/m3/src/gadgets/mul.rs

```diff
@@ -0,0 +1,408 @@
+// Copyright 2025 Irreducible Inc.
+
+use std::{array, marker::PhantomData};
+
+use anyhow::Result;
+use binius_field::{
+	ext_basis, packed::set_packed_slice, BinaryField, ExtensionField, Field, PackedExtension,
+	PackedField, TowerField,
+};
+
+use crate::builder::{
+	upcast_col, Col, Expr, TableBuilder, TableWitnessSegment, B1, B128, B32, B64,
+};
+
+/// Helper trait to create Multiplication gadgets for unsigned integers of different bit lengths.
+trait UnsignedMulPrimitives {
+	type FP: TowerField;
+	type FExpBase: TowerField + ExtensionField<Self::FP>;
+
+	const BIT_LENGTH: usize;
+
+	/// Computes the unsigned primitive mul of `x*y = z` and returns the tuple (z_high, z_low)
+	/// representing the high and low bits respectively.
+	fn mul(x: Self::FP, y: Self::FP) -> (Self::FP, Self::FP);
+
+	fn is_bit_set_at(a: Self::FP, index: usize) -> bool {
+		<Self::FP as ExtensionField<B1>>::get_base(&a, index) == B1::ONE
+	}
+
+	fn generator() -> Self::FExpBase;
+
+	/// Returns the generator shifted by the bit length of `Self::FP`.
+	fn shifted_generator() -> Self::FExpBase;
+}
+
+impl UnsignedMulPrimitives for u32 {
+	type FP = B32;
+	type FExpBase = B64;
+
+	const BIT_LENGTH: usize = 32;
+
+	fn mul(x: B32, y: B32) -> (B32, B32) {
+		let res = x.val() as u64 * y.val() as u64;
+		let low = B32::new(res as u32);
+		let high = B32::new((res >> 32) as u32);
+		(high, low)
+	}
+
+	fn generator() -> B64 {
+		B64::MULTIPLICATIVE_GENERATOR
+	}
+
+	fn shifted_generator() -> B64 {
+		let mut g = B64::MULTIPLICATIVE_GENERATOR;
+		for _ in 0..Self::BIT_LENGTH {
+			g *= g
+		}
+		g
+	}
+}
+
+impl UnsignedMulPrimitives for u64 {
+	type FP = B64;
+	type FExpBase = B128;
+
+	const BIT_LENGTH: usize = 64;
+
+	fn mul(x: B64, y: B64) -> (B64, B64) {
+		let res = x.val() as u128 * y.val() as u128;
+		let low = B64::new(res as u64);
+		let high = B64::new((res >> 64) as u64);
+		(high, low)
+	}
+
+	fn generator() -> Self::FExpBase {
+		B128::MULTIPLICATIVE_GENERATOR
+	}
+
+	fn shifted_generator() -> B128 {
+		let mut g = Self::generator();
+		for _ in 0..Self::BIT_LENGTH {
+			g *= g
+		}
+		g
+	}
+}
+
+// Internally used to have deduplicated implementations.
+#[derive(Debug)]
+struct Mul<
+	FExpBase: TowerField,
```

Comment:
These type parameters are redundant since they are associated types on `UX`. Please remove them from the `struct` type parameters, but you can keep them on the `impl` block for convenience.